src/components/lib/sl/sl.c

```diff
@@ -291,48 +306,101 @@ sl_thd_yield(thdid_t tid)
 	sl_thd_yield_cs_exit(tid);
 }
 
+static void
+sl_thd_aepinfo_init(struct sl_thd *t, thdcap_t thd, arcvcap_t rcv, tcap_t tc)
+{
+	assert(t);
+
+	sl_thd_aepinfo(t)->thd = thd;
+	sl_thd_aepinfo(t)->rcv = rcv;
+	sl_thd_aepinfo(t)->tc  = tc;
+}
+
 static struct sl_thd *
-sl_thd_alloc_init(thdid_t tid, thdcap_t thdcap)
+sl_thd_alloc_init(thdid_t tid, thdcap_t thdcap, arcvcap_t rcvcap, tcap_t tcap,
+		  asndcap_t sndcap, sl_thd_flags flags)
 {
 	struct sl_thd_policy *tp = NULL;
 	struct sl_thd        *t  = NULL;
 
-	tp             = sl_thd_alloc_backend(tid);
+	tp = sl_thd_alloc_backend(tid);
 	if (!tp) goto done;
-	t              = sl_mod_thd_get(tp);
-
-	t->thdid       = tid;
-	t->thdcap      = thdcap;
-	t->state       = SL_THD_RUNNABLE;
+	t  = sl_mod_thd_get(tp);
+	
+	t->thdid          = tid;
+	t->flags          = flags;
+	sl_thd_aepinfo_init(t, thdcap, rcvcap, tcap);
+	t->sndcap         = sndcap;
+	t->state          = SL_THD_RUNNABLE;
 	sl_thd_index_add_backend(sl_mod_thd_policy_get(t));
 
-	t->period      = t->timeout_cycs = t->periodic_cycs = 0;
-	t->wakeup_cycs = 0;
-	t->timeout_idx = -1;
-	t->prio        = TCAP_PRIO_MIN;
+	t->budget         = 0;
+	t->last_replenish = 0;
+	t->period         = t->timeout_cycs = t->periodic_cycs = 0;
+	t->wakeup_cycs    = 0;
+	t->timeout_idx    = -1;
+	t->prio           = TCAP_PRIO_MIN;
 
 done:
 	return t;
 }
 
-/* boot_thd = 1 if you want to create a boot-up thread in a separate component */
 static struct sl_thd *
-sl_thd_alloc_intern(cos_thd_fn_t fn, void *data, struct cos_defcompinfo *comp, int boot_thd)
+sl_thd_alloc_intern(cos_thd_fn_t fn, void *data)
 {
 	struct cos_defcompinfo *dci = cos_defcompinfo_curr_get();
 	struct cos_compinfo    *ci  = &dci->ci;
 	struct sl_thd          *t   = NULL;
 	thdcap_t thdcap;
-	thdid_t  tid;
+	thdid_t tid;
 
-	if (!boot_thd) thdcap = cos_thd_alloc(ci, ci->comp_cap, fn, data);
-	else           thdcap = cos_initthd_alloc(ci, comp->ci.comp_cap);
+	thdcap = cos_thd_alloc(ci, ci->comp_cap, fn, data);
 	if (!thdcap) goto done;
 
 	tid = cos_introspect(ci, thdcap, THD_GET_TID);
 	assert(tid);
-	t = sl_thd_alloc_init(tid, thdcap);
+	t = sl_thd_alloc_init(tid, thdcap, 0, 0, 0, 0);
 	sl_mod_thd_create(sl_mod_thd_policy_get(t));
+
+done:
+	return t;
+}
+
+static struct sl_thd *
+sl_thd_aep_alloc_intern(cos_aepthd_fn_t fn, void *data, struct cos_defcompinfo *comp, sl_thd_flags flags)
+{
+	struct cos_defcompinfo *dci = cos_defcompinfo_curr_get();
+	struct cos_compinfo    *ci  = &dci->ci;
+	struct sl_thd          *t   = NULL;
+	asndcap_t               snd = 0;
+	struct cos_aep_info     aep;
+	thdid_t                 tid;
+	int                     ret;
+
+	if (flags & SL_FLAG_SEND) {
+		struct cos_aep_info *sa;
+
+		assert(comp);
+		sa = cos_sched_aep_get(comp);
+
+		/* FIXME: inefficient.. find a zerocopy way! */
+		aep.tc  = sa->tc;
```

Comment:
I honestly think that `aep = *sa;` would work here (the compiler turns it into memcpy).

This whole approach is not pretty, but I think you did a good job working with the APIs we have.